File: src/challenge/1.platinumRift.cpp
#include <iostream>
#include <string>
#include <vector>
#include <algorithm>
#include <sstream>

using namespace std;

/* definition of the class Zone */
class Zone
{   
public:
    Zone(int zoneId, int platiniumSource);
    int getZoneId();
    int getPlatiniumSource();
    int getOwnerId();
    int* getTabNbPodEachPlayerPresent();
    vector<Zone> getAdjacentZones();
    vector<int> getVecIdAdjacentZones();
    int getNbOwnPodPresent(int myId);

    void setIdOwner(int idOwner);
    void addAdjacentZone(int zoneId);
    void addPodInTabPodEachPlayerPresent(int indexPlayer, int nbPod);
    void setTabPodEachPlayerPresent(int tabPod[]);

    bool isOwnPodPresent(int myId);
    int findIdAjacentZoneWithMaxPlatinum(int myId);
    int createMovingOrders_randomMove();

    /* for debug */
    void display();

private:
    int m_id;
    int m_ownerId;
    int m_platiniumSource;
    vector<int> m_vecIdAdjacentZones;   // should be deleted ??
    vector<Zone> m_vecAdjacentZones;
    int m_tabNbPodEachPlayerPresent[4];
};

class ComparePlatinumQuantity
{
    public:
        bool operator()(Zone zone1, Zone zone2)
        {
            return zone1.getPlatiniumSource() > zone2.getPlatiniumSource();
        }
};

/* global variables */
vector<Zone> vecZones;

/* declaration of functions */
Zone getZoneWithId(int idZone);
void displayAllZones(vector<Zone> vecZones);

/* implementation of the class Zone */
Zone::Zone(int zoneId, int platiniumSource)
{
    m_id = zoneId;
    m_platiniumSource = platiniumSource;
    m_ownerId = -1;
    for(int i = 0; i < 4; i++)
    {
        m_tabNbPodEachPlayerPresent[i] = 0;
    }
}

int Zone::getZoneId()
{
    return m_id;
}

int Zone::getPlatiniumSource()
{
    return m_platiniumSource;
}

int Zone::getOwnerId()
{
    return m_ownerId;
}

int* Zone::getTabNbPodEachPlayerPresent()
{
    return m_tabNbPodEachPlayerPresent;
}

vector<int> Zone::getVecIdAdjacentZones()
{
    return m_vecIdAdjacentZones;
}

vector<Zone> Zone::getAdjacentZones()
{
    return m_vecAdjacentZones;
}   

void Zone::setIdOwner(int idOwner)
{
    m_ownerId = idOwner;
}

void Zone::addAdjacentZone(int zoneId)
{
    m_vecIdAdjacentZones.push_back(zoneId);
    m_vecAdjacentZones.push_back(getZoneWithId(zoneId));
}

void Zone::addPodInTabPodEachPlayerPresent(int indexPlayer, int nbPod)
{
    m_tabNbPodEachPlayerPresent[indexPlayer] += nbPod;
}

void Zone::setTabPodEachPlayerPresent(int tabPod[])
{
    for(int i = 0; i < 4; i++)
    {
        m_tabNbPodEachPlayerPresent[i] = tabPod[i];
    }
}

bool Zone::isOwnPodPresent(int myId)
{
    return m_tabNbPodEachPlayerPresent[myId] != 0;
}

int Zone::getNbOwnPodPresent(int myId)
{
    return m_tabNbPodEachPlayerPresent[myId];
}

int Zone::findIdAjacentZoneWithMaxPlatinum(int myId)
{
    vector<Zone> vecZonesPotentialDestination;

    // cerr<<"idZone : "<<m_id<<endl;
    for(int i = 0; i < m_vecAdjacentZones.size(); i++)
    {
        if(getZoneWithId(m_vecAdjacentZones[i].getZoneId()).getOwnerId() != myId)
            vecZonesPotentialDestination.push_back(m_vecAdjacentZones[i]);
    }
    // displayAllZones(vecZonesPotentialDestination);
    if(vecZonesPotentialDestination.empty())    // all the adjacent zones are mine -> random move
    {
        return createMovingOrders_randomMove();
    }
    else
    {
        for(int i = 0; i < vecZonesPotentialDestination.size(); i++)
        {
            if(vecZonesPotentialDestination[i].getPlatiniumSource() == 0)
                vecZonesPotentialDestination.erase(vecZonesPotentialDestination.begin() + i);
        }
        sort(vecZonesPotentialDestination.begin(), vecZonesPotentialDestination.end(), ComparePlatinumQuantity());

        if(!vecZonesPotentialDestination.empty())
            return vecZonesPotentialDestination[0].getZoneId();
        else
            return createMovingOrders_randomMove();
    }
}

void Zone::display()
{
    cerr<<"============================================="<<endl;
    cerr<<"Zone "<<m_id<<" owned by "<<m_ownerId<<" produces "<<m_platiniumSource<<" platinum"<<endl;
    cerr<<"Troops present : P0 -> "<<m_tabNbPodEachPlayerPresent[0]<<" | P1 -> "<<m_tabNbPodEachPlayerPresent[1]<<" | P2 -> "<<
        m_tabNbPodEachPlayerPresent[2]<<" | P3 -> "<<m_tabNbPodEachPlayerPresent[3]<<endl;
    cerr<<"-------- adjacent zones -------"<<endl;
    for(int i = 0; i < m_vecIdAdjacentZones.size(); i++)
    {
        cerr<<m_vecIdAdjacentZones[i]<<" | ";
    }
    cerr<<endl;
}

int Zone::createMovingOrders_randomMove()
{
    int index;
    int nbOfDest = m_vecAdjacentZones.size();

    index = rand() % nbOfDest;

    return m_vecAdjacentZones[index].getZoneId(); 
}

/* function */
void displayAllZones(vector<Zone> vecZones)
{
    for(int i = 0; i < vecZones.size(); i++)
    {
        vecZones[i].display();
    }
}

Zone getZoneWithId(int idZone)
{
    for(int i = 0; i < vecZones.size(); i++)
    {
        if(vecZones[i].getZoneId() == idZone)
        {
            return vecZones[i];
            break;
        }
    }
}

string buildOrdersForOneMovement(int idZoneOrigine, int idZoneDest)
{
    string orders;
    ostringstream oss1;
    ostringstream oss2;

    oss1 << idZoneOrigine;
    orders += "1 " + oss1.str() + " "; // '1' for the move of 1 Pod 
    oss2 << idZoneDest;
    orders += oss2.str() + " ";

    return orders;
}

int createMovingOrders_attractPlatinumMove(Zone zone, int myId)
{
    return zone.findIdAjacentZoneWithMaxPlatinum(myId);
}

string createMovingOrders(int myId)
{
    int idZoneDest = 0;
    int idZoneOrigine = 0;
    // int nbOfDest = 0;
    int indexDest = 0;
    string orders;

    for(int i = 0; i < vecZones.size(); i++)
    {
        if(vecZones[i].isOwnPodPresent(myId))
        {
            idZoneOrigine = vecZones[i].getZoneId();
            // cerr<<"idZone : "<<idZoneOrigine<<" nbOfDest : "<<nbOfDest<<endl;
            // cerr<<"indexAlea : "<<indexAlea<<" idZoneDest : "<<idZoneDest<<endl;
            
            // idZoneDest = vecZones[i].createMovingOrders_randomMove();
            idZoneDest = createMovingOrders_attractPlatinumMove(vecZones[i], myId);
            vecZones[i].addPodInTabPodEachPlayerPresent(myId, 1);

            orders += buildOrdersForOneMovement(idZoneOrigine, idZoneDest);
        }
    }
    if(!orders.empty())
    {
        orders.pop_back();
        return orders;
    }
    else
        return "WAIT";
}

int findZoneOfDeployment(int myId)
{
    vector<int> vecIdZonesWithPlatinum;
    // displayAllZones(vecZones);
    for(int i = 0; i < vecZones.size(); i++)
    {
        if(vecZones[i].getPlatiniumSource() > 0)
        {
            vecIdZonesWithPlatinum.push_back(vecZones[i].getZoneId());
            // cerr<<"id zone with platinum "<<vecIdZonesWithPlatinum[i]<<endl;
        }
    }
    // for(int i = 0; i < vecIdZonesWithPlatinum.size(); i++)
    // {
    //     cerr<<"zone plat : "<<vecIdZonesWithPlatinum[i]<<endl;
    // }
    
    for(int i = 0; i < vecIdZonesWithPlatinum.size(); i++)
    {
        int zoneOwnerId = vecZones[i].getOwnerId();
        // cerr<<"zoneOwnerId : "<<zoneOwnerId<<endl;
        if((zoneOwnerId == myId || zoneOwnerId == -1) && (vecZones[i].getTabNbPodEachPlayerPresent()[myId] == 0))
        {
            vecZones[i].addPodInTabPodEachPlayerPresent(myId, 1);
            // displayAllZones(vecZones);
            return vecZones[i].getZoneId();
        }
    }
    return vecIdZonesWithPlatinum[0]; 
}

string createBuyingOrders(int myId, int platinumReserve)
{
    string orders;
    int deployingZoneId = 0;

    if(platinumReserve >= 20)
    {
        while(platinumReserve >= 20)
        {
            ostringstream oss;
            // cerr<<"platinumReserve : "<<platinumReserve<<endl;
            deployingZoneId = findZoneOfDeployment(myId);
            // displayAllZones(vecZones);
            // cerr<<"deployingZoneId : "<<deployingZoneId<<endl;
            oss << deployingZoneId;
            orders += "1 " + oss.str() + " ";
     
            platinumReserve -= 20;
        }
        orders.pop_back();
        return orders;
    }
    else
        return "WAIT";
}

int main()
{
    int playerCount; // the amount of players (2 to 4)
    int myId; // my player ID (0, 1, 2 or 3)
    int zoneCount; // the amount of zones on the map
    int linkCount; // the amount of links between all zones

    srand (time(NULL));
    cin >> playerCount >> myId >> zoneCount >> linkCount; cin.ignore();
    // cerr<<"zoneCount : "<<zoneCount<<endl;

    for (int i = 0; i < zoneCount; i++)
    {
        int zoneId; // this zone's ID (between 0 and zoneCount-1)
        int platinumSource; // the amount of Platinum this zone can provide per game turn
        cin >> zoneId >> platinumSource; cin.ignore();
        vecZones.push_back(Zone(zoneId, platinumSource));
    }

    for (int i = 0; i < linkCount; i++) {
        int zone1;
        int zone2;
        cin >> zone1 >> zone2; cin.ignore();
        vecZones[zone1].addAdjacentZone(zone2);
        vecZones[zone2].addAdjacentZone(zone1);
    }
    sort(vecZones.begin(), vecZones.end(), ComparePlatinumQuantity());

    // displayAllZones(vecZones);

    bool firstRound = true;
    // while(counter < 1) 
    while(1)
    {
        int platinumReserve;
        cin >> platinumReserve; cin.ignore();

        for (int i = 0; i < zoneCount; i++)
        {
            int zId; // this zone's ID
            int ownerId; // the player who owns this zone (-1 otherwise)
            int podsP0; // player 0's PODs on this zone
            int podsP1; // player 1's PODs on this zone
            int podsP2; // player 2's PODs on this zone (always 0 for a two player game)
            int podsP3; // player 3's PODs on this zone (always 0 for a two or three player game)
            cin >> zId >> ownerId >> podsP0 >> podsP1 >> podsP2 >> podsP3; cin.ignore();
            // cerr<<zId<<" | "<<ownerId<<" | "<<podsP0<<" | "<<podsP1<<" | "<<podsP2<<" | "<<podsP3<<endl;

            int tabPod[4] = {podsP0, podsP1, podsP2, podsP3};
            for(int j = 0; j < zoneCount; j++)
            {
                if(vecZones[j].getZoneId() == zId)
                {
                    // cerr<<"j : "<<j<<" vecZOne i : "<<vecZones[j].getZoneId()<<endl;
                    // cerr<<"ownerId : "<<ownerId<<endl;
                    vecZones[j].setIdOwner(ownerId);
                    vecZones[j].setTabPodEachPlayerPresent(tabPod);
                    break;
                }
            }       
        }

        // displayAllZones(vecZones);
        string moveOrderTemp = createMovingOrders(myId);
        string buyingAndDeploy = createBuyingOrders(myId, platinumReserve);

        cout << moveOrderTemp << endl; // movements
        if(firstRound)
        {
            cout<<"WAIT"<<endl;
            firstRound = false;
        }
        else
        {
            cout << buyingAndDeploy <<endl; // buying and deploy
        }

        //cerr<<"move : "<<moveOrderTemp<<endl;
        //cerr<<"buyAndDeploy : "<<buyingAndDeploy<<endl;
    }
}

File: src/challenge/scala/1.platinumRift.scala
import math._
import scala.util._
import scala.collection.mutable.ArrayBuffer
import scala.util.control.Breaks._
import java.util.Random

/**
 * TicToc allows for easy benchmarking of different parts in source code.
 * Just inherit the TicToc trait, then start a time measurement with tic,
 * and stop it using toc.
 * As of now tic-tocs can be nested but can not overlap, as the underlying
 * data structure is a stack: a toc always terminates the last tic.
 * The reason for this decision was that the TicToc trait and its methods
 * should be as lightweight as possible in order not to falsify the measured
 * run times.
 * @author fgysin
 */
trait TicToc {
  import scala.collection.mutable.LinkedList
  import scala.collection.mutable.Stack
  import java.io._

  var timeCounter: Int = 0
  var tics: Stack[Long] = Stack()
  var times: LinkedList[(String, Long)] = LinkedList()

  /**
   * Starts a time measurement.
   */
  def tic() {
    tics = tics.push(System.currentTimeMillis())
  }

  /**
   * Stops the last started time measurement.
   * @param descr a description of the last measurement period
   */
  def toc(descr: String) {
    var t = System.currentTimeMillis() - tics.pop()
    if (null != descr && !"".equals(descr))
      times = times :+ (descr, t)
    else {
      times = times :+ ("time" + timeCounter, t)
      timeCounter += 1
    }
  }

  /**
   * Write the logged times to a file. There is some formatting sugar contained
   * that checks if a file with this name already exists. If yes, the run times
   * are added at the bottom of the existing file (this is useful if you repeat
   * the same measurement multiple times), else the file is created.
   * @param path The path of the file to write the times log.
   */
  def writeTimesLog(path: String) {
    var linesToPrint = outLines
    try {
      // Check if the file was alreay created, if yes only
      // append the timings not the timing information.
      val lr = new LineNumberReader(new FileReader(path))
      val read = lr.readLine()
      if (null != read && read.startsWith("Timings")) {
        linesToPrint = linesToPrint.drop(2)
      }
      lr.close()
    } catch {
      case _ =>
    }
    // Write the timings and or timing information.
    val fw = new FileWriter(path, true);
    linesToPrint.foreach { l => fw.write(l + "\n"); }
    fw.close()
  }

  /**
   * Print the logged times.
   */
  def printTimesLog() {
    outLines.foreach { l => println(l); }
  }

  /**
   * String output of the current times and their respective descriptions.
   * @return
   */
  def outLines(): List[String] = {
    var list = List[String]()
    list ::= ("Timings of " + this.getClass().toString())
    list ::= "description:" + times.map(x => "\t" + x._1).reduce(_ + _)
    list ::= "times(ms):" + times.map(x => "\t" + x._2).reduce(_ + _)
    list.reverse
  }

}

class Node(val id: Int, val platinumSource: Int, var ownerId: Int, var pods: Array[Int]) {
    override def toString = f"Id : $id%d, platinumSource : $platinumSource%d, ownerId : $ownerId \n" + pods.mkString(", ")

    def update(ownerId: Int, pods: Array[Int]){
        this.ownerId = ownerId
        pods.copyToArray(this.pods)
    }

    def podsOnePlayerArePresent(idPlayer: Int): Boolean = if(pods(idPlayer) != 0) true else false
    def isZoneFree(myId: Int): Boolean = if(myId != this.ownerId) true else false
    def isZoneWithEnemy(myId: Int): Boolean = if(isZoneFree(myId) && ownerId != -1) true else false
    def isZoneWithPlatinum: Boolean = if(platinumSource > 0) true else false
    def ownedBy(ownerId: Int): Boolean = if(ownerId == this.ownerId) true else false
}

class Graph(val nbPlayers: Int, val myId: Int, val nbTotalZones: Int, val nbTotalLinks: Int) {
    var nodes: Map[Int, Node] = Map()
    var platinumZones: ArrayBuffer[Int] = ArrayBuffer()
    var adjacence: ArrayBuffer[ArrayBuffer[Int]] = ArrayBuffer()
    var myPlatinum = 0
    var idZoneMySpawn = 0
    var idZoneEnemySpawn = 0

    def isAllPlatinumZonesOwned: Boolean = {
        for(i <- platinumZones){
            if(!nodes(i).ownedBy(myId)) return false
        }
        return true
    }

    def idAdjacentZoneFree(idZone: Int): Int = {
        for(i <- adjacence(idZone)){
            if(nodes(i).ownerId != myId) return i
        }
        return -42
    }

    def isHalfPlatinumZonesOwned: Boolean = {
        var nbPlatinumZonesOwned = 0
        for(i <- platinumZones){
            if(nodes(i).ownerId == myId) nbPlatinumZonesOwned +=1
        }
        // Console.err.println("nbPlatinumZonesOwned : " + nbPlatinumZonesOwned)
        return (nbPlatinumZonesOwned >= (platinumZones.length / 3))
    }

    override def toString = {
        var listNode = ""
        for(i <- nodes){
            listNode += i.toString + "\n"
        }
        listNode
    }

    def displayNeighbours(idZone: Int) {
        Console.err.println(adjacence(idZone).mkString(", "))
    }

    private def widthGraphMove(idZoneFrom: Int): String = {
        var orders = ""
        var queue = new scala.collection.mutable.Queue[Int]
        var actualIdNode = 0
        var markedNode: ArrayBuffer[Int] = ArrayBuffer()
        var idTarget = 0
        var fathers: Map[Int, Int] = Map()
        var idZoneFather = 0
        var zoneConditionsToBeATarget = false

        if(nodes(idZoneFrom).ownedBy(myId) && nodes(idZoneFrom).podsOnePlayerArePresent(myId)){
            var idZoneFreeNextToActualZone = idAdjacentZoneFree(idZoneFrom)
            if(idZoneFreeNextToActualZone != -42) {
                idZoneFather = idZoneFreeNextToActualZone
            }
            else {
                queue += idZoneFrom
                breakable { while(!queue.isEmpty){
                    actualIdNode = queue.dequeue
                    markedNode += actualIdNode
                    if(!isHalfPlatinumZonesOwned){
                        zoneConditionsToBeATarget = nodes(actualIdNode).isZoneFree(myId) && nodes(actualIdNode).isZoneWithPlatinum
                    }
                    else{
                         zoneConditionsToBeATarget = nodes(actualIdNode).isZoneFree(myId)
                    }
                    if(zoneConditionsToBeATarget) {
                            // Console.err.println("Plus de platinum !!!")
                            idTarget = actualIdNode
                            idZoneFather = idTarget
                            while(fathers(idZoneFather) != idZoneFrom){
                                idZoneFather = fathers(idZoneFather)
                            }
                            // Console.err.println("idZoneFrom : " + idZoneFrom + " -> father : " + idZoneFather + " dest : " + idTarget)
                            break
                        }

                    for(i <- adjacence(actualIdNode)){
                        if(!markedNode.contains(i) && !queue.contains(i)){
                            queue += i
                            fathers += (i -> actualIdNode)
                        }
                    }
                }}
            }
            nodes(idZoneFather).ownerId = myId
            orders += "1 " + idZoneFrom + " " + idZoneFather + " "
        }
        orders
    }

    def move: String = {
        var orders = ""
        nodes.keys.foreach{ i =>
            breakable {while(nodes(i).pods(myId) > 0){
                // if(i == idZoneMySpawn && nodes(i).pods(myId) <= 5) break
                orders += widthGraphMove(i)
                if(!orders.isEmpty) nodes(i).pods(myId) -= 1
            }}
        }
        if(orders.length != 0) orders else "WAIT"
    }

    private def randomMove(idZoneFrom: Int): String = {
        var orders = ""
        if(nodes(idZoneFrom).ownedBy(myId) && nodes(idZoneFrom).podsOnePlayerArePresent(myId)){
            // var neighboursFree = for{i <- adjacence(idZoneFrom) if(nodes(i).isZoneFree(myId)} yield i
            var neighbours = adjacence(idZoneFrom)
            // Console.err.println("idZone : " + idZoneFrom + " -> " + neighbours.mkString(", "))
            val rand = new Random(System.currentTimeMillis());
            val random_index = rand.nextInt(neighbours.length);
            // Console.err.println("from : " + idZoneFrom + " to " + neighbours(random_index))
            orders += "1 " + idZoneFrom + " " + neighbours(random_index) + " "
        }
        orders
    }

    def invade: String = {
        var orders = ""

        // nodes.keys.foreach{ i =>
        //     if(myPlatinum >= 20){
        //         orders += spawnNextToFreeZone(i)
        //     }
        // }
        // if(orders.length != 0) orders else "WAIT"
        return "WAIT"
    }

    private def spawnNextToFreeZone(idZone: Int): String = {
        var orders = ""
        // Console.err.println(nodes(idZone).toString)
        if(nodes(idZone).ownedBy(myId)){
            var neighbours = adjacence(idZone)
            var neighboursWithEnemy = for{i <- neighbours if(nodes(i).isZoneWithEnemy(myId))}yield i
            if(neighboursWithEnemy.length != 0){
                breakable {for(j <- 0 until neighboursWithEnemy.length) if(nodes(neighboursWithEnemy(j)).isZoneFree(myId)) {
                    // Console.err.println("Zone %d is free", nodes(neighbours(j)).id)
                    orders += "1 " + idZone + " "
                    myPlatinum -= 20
                    // Console.err.println(orders)
                    break
                }}
            }
            // else
            // {
            //     // Console.err.println("idZone : " + idZone + " -> " + neighbours.mkString(", "))
            //     breakable {for(j <- 0 until neighbours.length) if(nodes(neighbours(j)).isZoneFree(myId)) {
            //         // Console.err.println("Zone %d is free", nodes(neighbours(j)).id)
            //         orders += "1 " + nodes(idZone).id + " "
            //         myPlatinum -= 20
            //         // Console.err.println(orders)
            //         break
            //     }}
            // }
        }

        orders
    }
}

object Player extends TicToc{

    def main(args: Array[String]) {
        // nbPlayers: the amount of players (2 to 4)
        // myId: my player ID (0, 1, 2 or 3)
        // nbZonesTotal: the amount of zones on the map
        // nbLinksTotal: the amount of links between all zones
        val Array(nbPlayers, myId, nbZonesTotal, nbLinksTotal) = for(i <- readLine split " ") yield i.toInt
        var graph = new Graph(nbPlayers, myId, nbZonesTotal, nbLinksTotal)

        for(i <- 0 until nbZonesTotal) {
            // zoneId: this zone's ID (between 0 and nbZonesTotal-1)
            // platinumSource: the amount of Platinum this zone can provide per game turn
            val Array(zoneId, platinumSource) = for(i <- readLine split " ") yield i.toInt
            graph.adjacence += ArrayBuffer()
            graph.nodes += (zoneId -> new Node(zoneId, platinumSource, -1, Array(0, 0, 0, 0)))
            if(platinumSource > 0) graph.platinumZones += zoneId
        }
        // Console.err.println(graph.platinumZones.mkString(", "))

        for(i <- 0 until nbLinksTotal) {
            val Array(zone1, zone2) = for(i <- readLine split " ") yield i.toInt
            graph.adjacence(zone1) += zone2
            graph.adjacence(zone2) += zone1
        }
        // graph.displayNeighbours(0)

        var firstTurn = true
        while(true) {
            val myPlatinum = readInt // my available Platinum
            graph.myPlatinum = myPlatinum
            for(i <- 0 until nbZonesTotal) {
                // zoneId: this zone's ID
                // ownerId: the player who owns this zone (-1 otherwise)
                // podsp0: player 0's PODs on this zone
                // podsp1: player 1's PODs on this zone
                // podsp2: player 2's PODs on this zone (always 0 for a two player game)
                // podsp3: player 3's PODs on this zone (always 0 for a two or three player game)
                val Array(zoneId, ownerId, podsp0, podsp1, podsp2, podsp3) = for(i <- readLine split " ") yield i.toInt
                graph.nodes(zoneId).update(ownerId, Array(podsp0, podsp1, podsp2, podsp3))
            }
            if(firstTurn) {
                for(i <- 0 until nbZonesTotal) {
                    if(graph.nodes(i).ownedBy(myId)) graph.idZoneMySpawn = i
                    if(graph.nodes(i).isZoneWithEnemy(myId)) graph.idZoneEnemySpawn = i
                }
                // Console.err.println("my spawn : " + graph.idZoneMySpawn)
                // Console.err.println("his spawn : " + graph.idZoneEnemySpawn)
                firstTurn = false
            }
            //Console.err.println(graph.toString)

            //tic
            println(graph.move) // first line for movement commands, second line for POD purchase (see the protocol in the statement for details)
            //toc("Move")
            //Console.err.println(outLines)
            println(graph.invade)
        }
    }
}
